### ISSUE TITLE
NVDA crashes Firefox when pressing shift+f10 twice in a text box

### ISSUE THREAD

**Issue by jcsteh** [nvaccess/nvda#8000]
Originally reported by @DerekRiemer.

### Steps to reproduce:

This is not reproduceable 100% of the time, but it shouldn't take too many repeats to trigger it; it certainly doesn't on my system.

1. Open this URL (though any text box in a web document should do): `data:text/html,<textarea>abcd</textarea>`
2. Tab to the text box and ensure NVDA is in focus mode.
3. Press shift+f10 twice.

### Expected behavior:
The first press should open the menu. The second should ideally do nothing.

### Actual behavior:
Firefox crashes!

### System configuration:
NVDA version:
next-14860,d6c75234

NVDA Installed or portable:
installed

Windows version:
Windows 10 Version 1709 (OS Build 17074.1002) 64 bit

Name and version of other software in use when reproducing the issue:
Firefox Nightly 60.0a1 (2018-02-14) (64-bit)

### Additional info:

The way I reproduce this here is somewhat contrived; most users wouldn' tpress shift+f10 twice. However, it's much easier to reproduce this way.

This also occurs if you press shift+f10 while Firefox happens to be blocked on a COM call to a content process, which happens during virtual buffer updates. This can certainly happen in the wild and is how @DerekRiemer sees it in normal usage. It's rare for him, but does happen. This seems to open the System menu, perhaps because Windows handles the message itself while it's in a COM modal loop.

[Firefox crash dump](https://crash-stats.mozilla.com/report/index/0a765699-4c9c-4c59-9dcf-c581c0180215)

We crash in `inProcess_getMessageHook` while handling an `execInThread` request. Here's what's happening:

1. NVDA calls `nvdaInProcUtils_IA2Text_findContentDescendant` via RPC.
2. That call arrives on an NVDA RPC thread within Firefox.
3. It calls `execInThread`, passing it a lambda. `execInThread` sends a window message to the main thread, passing the lambda, and then blocks, waiting until completion.
4. The main thread handles the `execInThread` request, runs the lambda and signals the RPC thread that it's done.
5. The RPC function returns.
6. For some reason, Firefox re-posts our `execInThread` window message (the one from step 3) to the main thread message queue.
7. The main thread tries to run the lambda passed in step 3, but it crashes because that lambda went out of scope in step 5.

Why Firefox re-posts the message is an open question. My guess is that it has something to do with the System menu opening. Perhaps that runs within a modal loop and consumed messages then get re-posted, whether by Firefox or by Windows itself.

While this is arguably a bug in Firefox or Windows, it's not an app's responsibility to protect itself against code we inject into its process, especially since we're using a message hook. Also, I suspect tracking down the cause in Firefox and fixing it is just not worth the effort. Finally, if it's happening in Firefox, it could conceivably happen in other apps.

We should be able to work around this by somehow guarding against processing the same `execInThread` request twice. This is probably going to require a global variable.

**Comment by Adriani90** [comment]
I can reproduce this reliably with NVDA 2017.4 and Windows 8.1 using Firefox 58.0.1.

**Comment by jcsteh** [comment]
Note that Firefox bringing up the System menu in the first place when you press shift+f10 in some cases is a bug, and without that bug, the circumstances that trigger this crash don't occur. That Firefox bug also causes other undesirable behaviour, so I ended up tracking this down in Firefox and submitting a patch. See [Mozilla bug 1382199](https://bugzilla.mozilla.org/show_bug.cgi?id=1382199).

This `execInThread` problem is still theoretically something NVDA should protect against, but if the System menu bug gets fixed in Firefox, it isn't necessary to fix `execInThread` from a practical standpoint.

**Comment by Brian1Gaff** [comment]
Is this the good old problem of using the mouse to close stuff and crashing 
nvda, when the keyboard route works properly? I always thought this was a 
windows bug.
 Brian

**Comment by jcsteh** [comment]
NO. The bug you speak of only exists in command consoles.

**Comment by zahra21** [comment]
hi.
did you try with previous versions of nvda and firefox?
today, i tested with nvda 2016.4 and firefox 51 and could not reproduce the problem.
i recommend that test with firefox esr or older versions (not quantum)
to see the result.
i said not quantum, because i think older versions of nvda not be compatible with firefox quantum.
and mentioned nvda 2017.1 or nvda 2016.4, because dont have the feature Replace execInWindow with execInThread
does realy replacing with exec in thread causes this crash?
is it a realy regression of this replacement?
does it make other issues too?
is it possible that the change that happenedto resolve the crash of firefox using multiprocess, makes this crash or even new crash, that maybe descover in the future?
 today with nvda 2016.4 i opened this page in firefox 51.
then tried to reproduce the crash many times.
i press shift f10, in the first time, context menu opened as expected.
in the second time, nvda said system subMenu.
in my second try, in the first time, nvda told context menu as normal,
 and in the second time, said the title of the page.
i tried many times and sometimes in the second try, nvda announced system subMenu and sometimes said the title of the page.

**Comment by jcsteh** [comment]
[**@zahra21**](https://github.com/zahra21) commented on [Feb 18, 2018, 2:14 AM GMT+10](https://github.com/nvaccess/nvda/issues/8000#issuecomment-366451569 "2018-02-17T16:14:02Z - Replied by Github Reply Comments"):
> does realy replacing with exec in thread causes this crash?

It'd be more accurate to say that the execInThread change creates the circumstances which allow this crash to happen, but in effect, yes, it's a regression introduced by this change. However, this change was necessary to support Firefox Quantum, so reverting it is not an option.

> does it make other issues too?

Not that we're aware of. That code has existed for quite some time now and this is the first (rare) issue we've discovered triggered by a very strange set of circumstances.

**Comment by jcsteh** [comment]
Even though the System menu bug is hopefully going to get fixed in Firefox, it occurred to me that we should be able to fix this trivially by just replacing the message with WM_NULL in the hook, since WH_GETMESSAGE hooks are able to modify the message. Strangely, this didn't fix the crash, even when I also set wParam to NULL. My only guess is that whatever mechanism is re-posting the messages must be using a WH_GETMESSAGE hook itself (or something equivalent) and it runs before ours (and thus before we've been able to mutate the message).

I do see some notes in PostThreadMessage about messages being lost in modal loops, so perhaps something is trying to compensate for that?

**Comment by MarcoZehe** [comment]
That’s perhaps because the bug was fixed on the Firefox side in version 60. ;)

**Comment by derekriemer** [comment]
Voting for a no close since other apps could have this.

**Comment by jcsteh** [comment]
Closing because we don't have a practical need for a fix on the NVDA side at this point (since it's fixed in Firefox). Also, the fix I tried didn't work (https://github.com/nvaccess/nvda/issues/8000#issuecomment-367832963), and whlie Mick and I did discuss some other possible fixes, they were all fairly ugly and might end up causing other problems. We can always reopen/file a new issue if we see this crop up again.
